### PR TITLE
[tests] Converting test_conv3d_extra into GTest

### PR DESCRIPTION
**Pull request by reidkwja**
GTests cases for 3D Convolution

### CODE REVIEW

File: test/gtest/conv3d_test.cpp [hunk 1]
```diff
@@ -67,19 +67,19 @@ void Run3dDriver(miopenDataType_t prec)
...
     case miopenInt8:
     case miopenInt32:
     case miopenDouble:
     case miopenFloat8:
     case miopenBFloat8:
-        FAIL() << "These data types are not supported by the 3D convolution test";
-        break;
+        FAIL() << "miopenHalf, miopenInt8, miopenBFloat16, miopenInt32, "
+                  "miopenDouble, miopenFloat8, miopenBFloat8 "
+                  "data type not supported by conv_igemm_dynamic test";
```
Comment (xinlipn) [inline test/gtest/conv3d_test.cpp]: > Change conv_igemm_dynamic to conv2d_test maybe?

File: test/gtest/conv3d_test.cpp [hunk 2]
```diff
@@ -128,19 +128,19 @@ std::vector<TestCase> GetTestCases(const std::string& precision)
     const std::vector<TestCase> test_cases = {
```
Comment (xinlipn) [inline test/gtest/conv3d_test.cpp]: std::vector\<TestCase\> wont be necessary for this case, you can use std::vector\<std::string\> instead. Here's an example

test/gtest/conv_hip_igemm_xdlops.cpp

**Comment by xinlipn** [comment]
@reidkwja  Maybe a good idea to merge with this PR below, which has critical fixes for environment variable updates, e.g. changing MIOPEN_DECLARE_ENV_VAR to MIOPEN_DECLARE_ENV_VAR_BOOL

https://github.com/ROCmSoftwarePlatform/MIOpen/pull/2514

Here's an example FYI:

https://github.com/ROCmSoftwarePlatform/MIOpen/blob/589b8d12c4ea1426366284e41ebbcdada6bcdb2b/src/solver/conv_bin_wino3x3U.cpp#L38

https://github.com/ROCmSoftwarePlatform/MIOpen/blob/589b8d12c4ea1426366284e41ebbcdada6bcdb2b/src/solver/conv_bin_wino3x3U.cpp#L49

File: test/gtest/conv3d_test.cpp [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+ *
+ *******************************************************************************/
+#include <tuple>
+#include <miopen/miopen.h>
+#include <miopen/env.hpp>
+#include <gtest/gtest.h>
+#include "../conv3d.cpp"
+#include "get_handle.hpp"
+
+MIOPEN_DECLARE_ENV_VAR(MIOPEN_TEST_FLOAT_ARG)
```
Comment (reidkwja) [inline test/gtest/conv3d_test.cpp]: @xinlipn would this ENV be the right choice for 3DConv?

File: test/gtest/conv3d_test.cpp [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+        Run3dDriver(miopenFloat);
+    }
+    else
+    {
+        GTEST_SKIP();
+    }
+#endif
+};
+
+INSTANTIATE_TEST_SUITE_P(Conv3dTest, ConfigWithFloat, testing::ValuesIn(GetTestCases("--float")));
```
Comment (atamazov) [inline test/gtest/conv3d_test.cpp]: This should be named `Conv3dTestExtra`

File: test/gtest/conv3d_test.cpp [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+MIOPEN_DECLARE_ENV_VAR_BOOL(MIOPEN_TEST_COMPOSABLEKERNEL)
+MIOPEN_DECLARE_ENV_VAR_BOOL(MIOPEN_TEST_ALL)
+
+namespace conv3d_test {
+
+static bool IsTestRunWith(const char* float_arg)
+{
+    assert(float_arg != nullptr);
+    const auto& s_envVar = miopen::GetStringEnv(ENV(MIOPEN_TEST_FLOAT_ARG));
+    return (!s_envVar.empty() && std::strcmp(s_envVar.c_str(), float_arg) == 0);
```
Comment (atamazov) [inline test/gtest/conv3d_test.cpp]: The test should be also run when both `MIOPEN_TEST_ALL` and `MIOPEN_TEST_FLOAT_ARG` are not defined.

File: test/gtest/conv3d_test.cpp [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+
+TEST_P(ConfigWithFloat, FloatTest_conv3d_test)
+{
+#if MIOPEN_BACKEND_OPENCL
+
+    GTEST_SKIP() << "MIOPEN_BACKEND_HIP needed for this test";
+
+#else // MIOPEN_BACKEND_HIP, OCL_DISABLED
+    const auto& handle = get_handle();
+    if(IsTestSupportedForDevice(handle) && miopen::IsEnabled(ENV(MIOPEN_TEST_COMPOSABLEKERNEL)) &&
```
Comment (atamazov) [inline test/gtest/conv3d_test.cpp]: The check of `MIOPEN_TEST_COMPOSABLEKERNEL` should be removed.

File: test/gtest/conv3d_test.cpp [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+    return test_cases;
+}
+
+} // namespace conv3d_test
+
+using namespace conv3d_test;
+
+TEST_P(ConfigWithFloat, FloatTest_conv3d_test)
+{
+#if MIOPEN_BACKEND_OPENCL
```
Comment (atamazov) [inline test/gtest/conv3d_test.cpp]: [Q] Why we need to check MIOPEN_BACKEND_OPENCL? We do not maintain this backend anymore.

File: test/gtest/conv3d_test.cpp [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+        test_drive<conv3d_driver>(ptrs.size(), ptrs.data());
+        auto capture = testing::internal::GetCapturedStderr();
+        std::cout << capture;
+    }
+};
+
+bool IsTestSupportedForDevice(const miopen::Handle& handle)
+{
+    std::string devName = handle.GetDeviceName();
+    if(devName == "gfx94" || devName == "gfx103" || devName == "gfx110")
```
Comment (atamazov) [inline test/gtest/conv3d_test.cpp]: This test should be run on all devices.

File: test/gtest/conv3d_test.cpp [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+    case miopenFloat: params = ConfigWithFloat::GetParam(); break;
+    case miopenInt8:
+    case miopenHalf:
+    case miopenBFloat16:
+    case miopenInt32:
+    case miopenDouble:
+        FAIL() << "miopenHalf, miopenBFloat16, miopenInt8, miopenInt32, "
+                  "miopenDouble data "
+                  "type not supported by "
+                  "test_conv_hip_igemm_xdlops test";
```
Comment (atamazov) [inline test/gtest/conv3d_test.cpp]: The name of the test is wrong.

**Review by atamazov — CHANGES_REQUESTED** [review]

**Comment by atamazov** [comment]
The title should look like "[tests] Converting test_conv3d_extra into GTest". Please also mark this with https://github.com/ROCm/MIOpen/labels/testing.

File: test/gtest/conv3d_test.cpp (hunk 9 above)
Comment (atamazov) [reply test/gtest/conv3d_test.cpp]: [resolved]

File: test/gtest/conv3d_test.cpp (hunk 8 above)
Comment (atamazov) [reply test/gtest/conv3d_test.cpp]: [resolved] but I recommend removing the check of the device.

File: test/gtest/conv3d_test.cpp (hunk 7 above)
Comment (atamazov) [reply test/gtest/conv3d_test.cpp]: [resolved]

File: test/gtest/conv3d_test.cpp (hunk 6 above)
Comment (atamazov) [reply test/gtest/conv3d_test.cpp]: [resolved]

File: test/gtest/conv3d_test.cpp (hunk 4 above)
Comment (atamazov) [reply test/gtest/conv3d_test.cpp]: [resolved]

**Comment by junliume** [comment]
@xinlipn @atamazov re-request review

**Comment by xinlipn** [comment]
> @xinlipn @atamazov re-request review

@junliume , I have updated the code per @atamazov 's comments, may need him revisit the changes.

**Comment by xinlipn** [comment]
@atamazov , could you help review my latest changes when you have a chance?

File: test/gtest/conv3d_test.cpp [hunk 10]
```diff
@@ -40,7 +40,9 @@ static bool IsTestRunWith(const char* float_arg)
 {
     assert(float_arg != nullptr);
     const auto& s_envVar = miopen::GetStringEnv(ENV(MIOPEN_TEST_FLOAT_ARG));
-    return (!s_envVar.empty() && std::strcmp(s_envVar.c_str(), float_arg) == 0);
+    return ((!s_envVar.empty() && std::strcmp(s_envVar.c_str(), float_arg) == 0 &&
```
Comment (atamazov) [inline test/gtest/conv3d_test.cpp]: [Quality] I suspect this line is equivalent to
```C++
    return ((s_envVar == float_arg &&
```

File: test/gtest/conv3d_test.cpp [hunk 11]
```diff
@@ -40,7 +40,9 @@ static bool IsTestRunWith(const char* float_arg)
...
+             miopen::IsEnabled(ENV(MIOPEN_TEST_ALL))) ||
+            (s_envVar.empty() && miopen::IsUnset(ENV(MIOPEN_TEST_ALL))));
```
Comment (atamazov) [inline test/gtest/conv3d_test.cpp]: [Recommendation] I think this design looks cleaner:
```suggestion
    if(miopen::IsUnset(ENV(MIOPEN_TEST_ALL)))
        return true; // standalone run
    return miopen::IsEnabled(ENV(MIOPEN_TEST_ALL)) //
        && miopen::GetStringEnv(ENV(MIOPEN_TEST_FLOAT_ARG)) == float_arg;
```
line 42 is not needed anymore.

**Review by atamazov — CHANGES_REQUESTED** [review]

File: test/gtest/conv3d_test.cpp (hunk 5 above)
Comment (atamazov) [reply test/gtest/conv3d_test.cpp]: [resolved] but please see recent updates at https://github.com/ROCm/MIOpen/issues/1843#issue-1420976061

**Comment by atamazov** [comment]
@xinlipn 

> @atamazov , could you help review my latest changes when you have a chance?

Sure. I think this can be merged in, but I recommend making changes suggested at https://github.com/ROCm/MIOpen/pull/2554#discussion_r1464106309 first.

File: test/gtest/conv3d_test.cpp (hunk 11 above)
Comment (atamazov) [reply test/gtest/conv3d_test.cpp]: @xinlipn Please see recently updated https://github.com/ROCm/MIOpen/issues/1843#issue-1420976061 for more info.

**Review by junliume — APPROVED** [review]

**Pull request merged**